src/codegen/codegen-expression.lisp

```diff
@@ -289,7 +289,8 @@
 
 (defun data-letrec-able-p (initform env)
   (let ((ctor-ent (find-constructor initform env)))
-    (and ctor-ent
+    (and (node-direct-application-p initform)
```

Comment:
The implementation of recursive data requires a direct application:
https://github.com/coalton-lang/coalton/commit/11968b8180aaacc8710a76e91e1e48bfe7a7232e